### ISSUE TITLE
only last tap registered when tapping rapidly

### ISSUE THREAD

**Issue by NANASHI0X74** [gujjwal00/avnc#224]
I noticed that when I rapidly tap things, the server only registers one click, even if I tap on two locations on screen. Is this intentional and if yes is it configurable in any way?

**Comment by NANASHI0X74** [comment]
nixOS, sway desktop, wayvnc server

although I don't have this problem with other vnc clients when I use tigervnc on my laptop and rapidly click there

**Comment by NANASHI0X74** [comment]
https://github.com/gujjwal00/avnc/assets/17546398/508387ea-fd3a-40a2-986a-22b621868600

Here, I recorded an example

**Comment by gujjwal00** [comment]
I am unable to reproduce the problem on my side. 
Does this issue happen with all apps/games running on remote server?

Please try enabling delayed click events: Long press on the server => Edit => Advanced => Enable **Send delayed click events**

**Comment by gujjwal00** [comment]
Any updates @NANASHI0X74 ?

**Comment by NANASHI0X74** [comment]
Hi, yes- I just had the same behavior with a graphics program called pinta:

https://github.com/gujjwal00/avnc/assets/17546398/b0e38aaf-02dd-44d1-965b-dbdb2ce8c204

The behavior is the same with delayed click events enabled or disabled, I couldn't really notice a difference.

**Comment by gujjwal00** [comment]
Can you please post a screenshot of **Input** Settings in AVNC?

**Comment by gujjwal00** [comment]
One more thing: please go to AVNC Settings => Tools => Touch test. Tap in the grey area, just like you are doing in previous video with pinta. Copy & paste the logs here. Thanks.

**Comment by gujjwal00** [comment]
Never mind, I was able to reproduce it with pinta. Although, the root cause may the way gesture detection works in Android. Let me investigate a bit more.

**Comment by gujjwal00** [comment]
As far as I can tell, there are two cases here:
1. Rapid taps at/near single point - In this case 2 consecutive taps will be treated as double-tap by Android's gesture detector. If you tap 6 times, there will be 3 double-taps. If you tap 7 times, there will be 3 double-taps followed by 1 single-tap. For double-taps, AVNC will send a double-click event (or whatever you have set in settings).
2. Rapid taps at different points: In this case, taps aren't happening as same point, so gesture detector will not generate double-taps. But it will also not generate single-taps if taps are happening quicker than a threshold (~ 300 millisecond, known as double-tap timeout). So only the last tap will trigger single-tap.

To test this, I ran `xev | grep -n ButtonPress` but I don't know if that works under Wayland. You can sort of use pinta to test this theory: If you zoom-in in pinta to make touches larger, you can see that touch caused by rapid tapping at some point is darker than what is caused by single tap. You can also test the second case by rapidly tapping by both thumbs at far ends of the screen. Only the last tap will be registered.

Unfortunately, this is not something I can fix. The timeout is necessary for gesture detection to work.

**Comment by NANASHI0X74** [comment]
ahh, that does make sense.
🤔
is it worth considering adding an option in the input settings->double tap to turn off detection for that gesture or to make the "None" Option behave that way?

**Comment by gujjwal00** [comment]
That option only controls what happens **in response to** double-tap. The detection is done by Android framework APIs.

**Comment by mttcr** [comment]
Hello @gujjwal00 !
I am facing a similar issue in my company which uses aVNC to connect to remote servers. the vnc shows a browser, and every touch in the browser (touch in a input to insert text, touch on a button in the html page...) is slow:
- a single touch (tap) takes around half a second to get handled
- when typing rapidly in html buttons, half of the events are lost.

I tried with a bluetooth mouse and spammed clicks on the same buttons, it works like a charm.

It looks like this issue, events aren't propagated as-is from android api.
I dig a bit on the internet, and I found these links:
https://developer.android.com/reference/android/view/GestureDetector#setOnDoubleTapListener(android.view.GestureDetector.OnDoubleTapListener)
I saw  that you already handle setIsLongpressEnabled in your code: https://github.com/gujjwal00/avnc/blob/433092d8ed51ed45014f012e0af15938c54498d4/app/src/main/java/com/gaurav/avnc/ui/vnc/TouchHandler.kt#L369
maybe depending on the app configuration (disable all "double tap" events, etc.) we could also call setOnDoubleTapListener(null) ?

this stack overflow issue talks about onSingleTapUp instead of onSingleTapConfirmed  https://stackoverflow.com/questions/10824388/either-disable-double-tap-or-detect-second-tap-location

To sum up, if we add a custom option or depending if no double/long actions are configured, could we be faster for interpreting tap events? Depending on the complexity, my company could sponsors you a bit.
Thank you!
Mathieu.

**Comment by gujjwal00** [comment]
Hi @mttcr, excellent writup! That is indeed the whole story behind this issue. And I actually tried to implement the `onSingleTapUp()` optimization couple years back (I still have that "WIP" commit stashed somewhere). My aim was to remove this extra 300ms latency if separate double tap detection was not needed. 

> I tried with a bluetooth mouse and spammed clicks on the same buttons, it works like a charm.

Yeah, a mouse will give you the best experience because mouse events are immediately sent to the server.

> Depending on the complexity, my company could sponsors you a bit.

This feature is something I like so it will be implemented eventually. But any sponsorship is always appreciated, and if you or your company find AVNC useful, you are welcome to sponsor it.

**Comment by gujjwal00** [comment]
I have added support for this ([Test APK](https://github.com/gujjwal00/avnc/actions/runs/19101555110/artifacts/4473758960))

To allow rapid tapping:
- Keep 'Double tap' option set to `Double-click` (default value)
- Set 'Double tap and swipe' to `None` 

It will be available in next version.
